File: iosApp/iosApp/Posts/PostViewModel.swift
import Foundation
import Shared

@MainActor
class PostViewModel: ObservableObject {
    @Published var posts: [PostModel] = []
    @Published var isLoading = false
    @Published var errorMessage: String? = nil

    private let repository: PostRepository

    init(repository: PostRepository = PostRepository(apiClient: ApiClient().client)) {
        self.repository = repository
    }

    func fetchPosts() {
        isLoading = true
        errorMessage = nil

        Task {
            do {
                let result = try await self.repository.loadPosts()
                self.posts = result
            } catch {
                self.errorMessage = error.localizedDescription
            }
            self.isLoading = false
        }
    }
}

File: composeApp/src/commonMain/kotlin/id/my/komangsidhiartha/multipartuiapp/posts/PostViewModel.kt
package id.my.komangsidhiartha.multipartuiapp.posts

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

class PostViewModel(
    private val repository: PostRepository = PostRepository()
) : ViewModel() {
    private val _posts = MutableStateFlow<List<PostModel>>(emptyList())
    val posts: StateFlow<List<PostModel>> = _posts.asStateFlow()

    private val _isLoading = MutableStateFlow<Boolean>(false)
    val isLoading: StateFlow<Boolean> = _isLoading.asStateFlow()

    init {
        fetchPosts()
    }

    fun fetchPosts() {
        viewModelScope.launch {
            if (_isLoading.value) return@launch

            _isLoading.value = true
            try {
                _posts.value = repository.loadPosts()
            } finally {
                _isLoading.value = false
            }
        }
    }
}